src/generate_pisa_index.cpp

```diff
@@ -40,26 +21,34 @@ std::ostream &write_sequence(std::ostream &os, gsl::span<T> sequence)
     return os;
 }
 
-
 struct inverted_index {
 
-  std::unordered_map<uint32_t, std::vector<uint32_t>> m_documents {};
-  std::unordered_map<uint32_t, std::vector<uint32_t>> m_frequencies {};
-  std::map<uint32_t, uint32_t> m_document_sizes {};
+  std::vector<std::vector<uint32_t>> m_documents {};
+  std::vector<std::vector<uint32_t>> m_frequencies {};
+  std::vector<uint32_t> m_document_sizes {};
   std::vector<std::string> m_plain_terms {};
  
-  void add_document_lengths(const std::string doclen_file) {
+  void add_document_lengths(const std::string& doclen_file) {
       std::ifstream input_lengths(doclen_file);
       uint32_t doc_id;
       uint32_t size;
+      uint32_t expected_id = 0;
       while (input_lengths >> doc_id >> size) {
-          m_document_sizes[doc_id] = size;
+          if (doc_id != expected_id) {
+              std::cerr << "Document Length file needs to be sorted. Exiting."
+                        << std::endl;
+              exit(EXIT_FAILURE);
+          }
+          m_document_sizes.emplace_back(size);
```

Comment:
I would refrain from using `emplace` when passing the exact type of the element in the container. Effectively, you're calling a copy constructor anyway, and it might be suggesting that there's something more complicated going on.